### PR TITLE
tokio-sync: AtomicTask performance improved if task is current

### PR DESCRIPTION
**Pull request by seanmonstar**
This adds some more benchmarks to allow better measuring all parts of the channels, and then with the benchmarks guiding, improves performance of polls when the task would already notify the current.

On my machine:

Before:

```
test legacy::bounded_rx_not_ready      ... bench:         137 ns/iter(+/- 31)
test legacy::bounded_tx_poll_not_ready ... bench:         122 ns/iter(+/- 70)
test legacy::unbounded_rx_not_ready    ... bench:         136 ns/iter(+/- 14)
test legacy::unbounded_rx_not_ready_x5 ... bench:         473 ns/iter(+/- 24)
test tokio::bounded_rx_not_ready       ... bench:         121 ns/iter(+/- 3)
test tokio::bounded_tx_poll_not_ready  ... bench:         127 ns/iter(+/- 1)
test tokio::unbounded_rx_not_ready     ... bench:         149 ns/iter(+/- 2)
test tokio::unbounded_rx_not_ready_x5  ... bench:         440 ns/iter(+/- 25)
```

**After**:

```
test legacy::bounded_rx_not_ready      ... bench:         137 ns/iter(+/- 31)
test legacy::bounded_tx_poll_not_ready ... bench:         122 ns/iter(+/- 6)
test legacy::unbounded_rx_not_ready    ... bench:         135 ns/iter(+/- 9)
test legacy::unbounded_rx_not_ready_x5 ... bench:         473 ns/iter(+/- 67)
test tokio::bounded_rx_not_ready       ... bench:         105 ns/iter(+/- 7)
test tokio::bounded_tx_poll_not_ready  ... bench:         114 ns/iter(+/- 4)
test tokio::unbounded_rx_not_ready     ... bench:         104 ns/iter(+/- 4)
test tokio::unbounded_rx_not_ready_x5  ... bench:         346 ns/iter(+/- 6)
```

### CODE REVIEW

File: tokio-sync/src/task/atomic_task.rs [hunk 1]
```diff
@@ -168,12 +168,23 @@ impl AtomicTask {
...
+    fn do_register<F: FnOnce() -> Task>(&self, func: F) {
         debug!(" + register_task");
         match self.state.compare_and_swap(WAITING, REGISTERING, Acquire) {
             WAITING => {
                 unsafe {
                     // Locked acquired, update the waker cell
-                    self.task.with_mut(|t| *t = Some(task));
+                    self.task.with_mut(|t| {
+                        let should_update = (&*t).as_ref()
+                            .map(|prev| !prev.will_notify_current())
```
Comment (carllerche) [inline tokio-sync/src/task/atomic_task.rs]: I don't think this is correct. `register_task` could register a task that is *not* current.

There are a few options, but I'm not sure what is best.

Comment (seanmonstar) [reply tokio-sync/src/task/atomic_task.rs]: You mean that as the API is now, someone *could* call `atomic_task.register_task(some_other_task_because_reasons)`, and in the case, it should definitely register the task even if the current one would notify?

Comment (carllerche) [reply tokio-sync/src/task/atomic_task.rs]: Yep. I don’t remember why the api was added. It would be worth tracking down usage in Tokio.

Comment (seanmonstar) [reply tokio-sync/src/task/atomic_task.rs]: I've updated from a simple `FnOnce` generic to a more specialized trait, only check for `will_notify_current` in the `atomic.register()` case.

**Comment by carllerche** [comment]
We should add a test that checks the `register_task(not_current)` issue.

**Pull request merged**